packages/react-reconciler/src/ReactFiberClassComponent.js

```diff
@@ -44,16 +44,24 @@ let didWarnAboutStateAssignmentForComponent;
 let warnOnInvalidCallback;
 
 if (__DEV__) {
+  const didWarnOnInvalidCallback = {};
   didWarnAboutStateAssignmentForComponent = {};
 
   warnOnInvalidCallback = function(callback: mixed, callerName: string) {
-    warning(
-      callback === null || typeof callback === 'function',
-      '%s(...): Expected the last optional `callback` argument to be a ' +
-        'function. Instead received: %s.',
-      callerName,
-      callback,
-    );
+    if (callback === null || typeof callback === 'function') {
+      return;
+    }
+    const key = `${callerName}_${JSON.stringify(callback)}`;
```

Comment:
Why does this need `JSON.stringify`? Just coercing to string seems enough to me.

Reply:
@gaearon If we remove `JSON.stringify`, `{foo: 'bar'}` and `new Foo()` will be coerced to the same string `[object Object]`. Shall we just remove the `toWarnDev` test cases about `new Foo()`?

Reply:
Yeah I think that's okay. Then maybe you don't even need to add new lines. Just add a comment where it's being deduplicated.

Reply:
@gaearon Seems all things are done in the new commit :)